contracts/origination/OriginationControllerSTIRFRY.sol

```diff
@@ -0,0 +1,268 @@
+
+            _validateCounterparties(signingCounterparty, callingCounterparty, msg.sender, externalSigner, sig, sighash);
+
+            // consume signer nonce
+            loanCore.consumeNonce(externalSigner, sigProperties.nonce, sigProperties.maxUses);
+        }
+
+        // initialize loan
+        loanId = _initialize(loanTerms, stirfryData, vaultAddress, borrower, lender);
+
+        // Run predicates check at the end of the function, after vault is in escrow. This makes sure
+        // that re-entrancy was not employed to withdraw collateral after the predicates check occurs.
+        if (itemPredicates.length > 0) originationHelpers.runPredicatesCheck(borrower, lender, loanTerms, itemPredicates);
+    }
+
+    // =========================================== HELPERS ==============================================
+
+    /**
+     * @notice Validate that the stirfry data provided is valid for the given loan terms.
+     *
+     * @param loanTerms                     The terms of the loan.
+     * @param stirfryData                   The stirfry data to validate.
+     * @param vaultAddress                  The address of the vault.
+     */
+    function _validateStirfryTerms(
+        LoanLibrary.LoanTerms calldata loanTerms,
+        StirfryData calldata stirfryData,
+        address vaultAddress
+    ) internal view {
+        // verify the vaulted currency is allowed to be paired with the loan terms payable currency
+        bytes32 currencyHash = keccak256(
+            abi.encodePacked(
+                loanTerms.payableCurrency,
+                stirfryData.vaultedCurrency,
+                stirfryData.payableToVaultedCurrencyRatio
+            )
+        );
+        if(stirfryPairs[currencyHash] != true)
+            revert OCS_InvalidStirfryPair(loanTerms.payableCurrency, stirfryData.vaultedCurrency);
+
+        // verify the vaulted currency amounts
+        if(loanTerms.principal * stirfryData.payableToVaultedCurrencyRatio != stirfryData.lenderVaultedCurrencyAmount)
+            revert OCS_InvalidPrincipalAmounts(
+                loanTerms.principal,
+                stirfryData.payableToVaultedCurrencyRatio,
+                stirfryData.lenderVaultedCurrencyAmount
+            );
+
+        // verify that the vault holds the lenderVaultedCurrencyAmount
+        uint256 vaultAmount = IERC20(stirfryData.vaultedCurrency).balanceOf(vaultAddress);
+        if(vaultAmount != stirfryData.lenderVaultedCurrencyAmount)
+            revert OCS_InvalidVaultAmount(vaultAmount, stirfryData.lenderVaultedCurrencyAmount);
+
+        // calculate the total interest due over the loan duration
+        uint256 totalInterest = loanTerms.principal * loanTerms.durationSecs * loanTerms.interestRate
+            / (Constants.BASIS_POINTS_DENOMINATOR * Constants.SECONDS_IN_YEAR);
+
+        // verify interest amounts
+        if(totalInterest * stirfryData.payableToVaultedCurrencyRatio != stirfryData.borrowerVaultedCurrencyAmount)
```

Comment:
All these checks against the ratios seem to suggest that, if the ratio needs to match exactly, then there is no need for the `StirfryData` fields? See comment below on line 236